picard/ui/searchdialog.py

```diff
@@ -0,0 +1,466 @@
+        if table_loaded:
+            self.save_table_header_state()
+        config.persist["searchdialog_window_size"] = self.size()
+        self.search_box.save_checkbox_state()
+
+    def save_table_header_state(self):
+        state = self.table.horizontalHeader().saveState()
+        config.persist["searchdialog_header_state"] = state
+
+
+class TrackSearchDialog(SearchDialog):
+
+    def __init__(self, parent):
+        super(TrackSearchDialog, self).__init__(parent)
+        self.file_ = None
+        self.setWindowTitle(_("Track Search Results"))
+        self.table_headers = [
+                _("Name"),
+                _("Length"),
+                _("Artist"),
+                _("Release"),
+                _("Date"),
+                _("Country"),
+                _("Type")
+        ]
+
+    def search(self, text):
+        """Performs search using query provided by the user."""
+        self.retry_params = (self.search, text)
+        self.search_box.search_edit.setText(text)
+        self.show_progress()
+        self.tagger.xmlws.find_tracks(self.handle_reply,
+                query=text,
+                search=True,
+                limit=25)
+
+    def load_similar_tracks(self, file_):
+        """Performs search by using existing metadata information
+        from the file."""
+        self.retry_params = (self.load_similar_tracks, file_)
+        self.file_ = file_
+        metadata = file_.orig_metadata
+        query = {
+                'track': metadata['title'],
+                'artist': metadata['artist'],
+                'release': metadata['album'],
+                'tnum': metadata['tracknumber'],
+                'tracks': metadata['totaltracks'],
+                'qdur': str(metadata.length / 2000),
+                'isrc': metadata['isrc'],
+        }
+        if config.setting["use_adv_search_syntax"]:
+            # Display the query in advance syntax format.
+            query_str = ' '.join(['%s:(%s)' % (item, value) for item, value in query.iteritems() if value])
+        else:
+            # Display only the track title
+            query_str = query["track"]
+        # `query_str` is used only for presenting purpose. Actual query consists of all filters and follows
+        # advanced query syntax.
+        query["limit"] = 25
+        self.search_box.search_edit.setText(query_str)
+        self.show_progress()
+        self.tagger.xmlws.find_tracks(
+                self.handle_reply,
+                **query)
+
+    def retry(self):
+        """Retries the search using information from `retry_params`.
+
+        `retry_params` is a tuple having search information.
+        retry_params[0] -- Method to be used to perform search
+                        -- Can be `self.search()` or `self.load_similar_tracks()`
+        retry_params[1] -- Search query information
+                        -- Can be a text string, or a File object
+        """
+        self.retry_params[0](self.retry_params[1])
```

Comment:
Please use a [namedtuple](https://docs.python.org/2/library/collections.html?highlight=namedtuple#collections.namedtuple) instead of magic tuples (we already do so in [picard.script](https://github.com/metabrainz/picard/blob/master/picard/script.py#L72-L75)).